prov/tcp/src/tcpx_progress.c

```diff
@@ -209,19 +215,178 @@ static void progress_tx_queue(struct tcpx_ep *ep)
+
+	if (ret < 0) {
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,
+			"fi_epoll_wait failed\n");
+		fastlock_release(&mgr->lock);
+		return -errno;
+	}
 
-	ep = container_of(util_ep, struct tcpx_ep, util_ep);
+	for (i = 0; i < ret ; i++) {
+		ep = (struct tcpx_ep *) mgr->ctxs[i];
+		progress_rx_queue(ep);
+	}
+	fastlock_release(&mgr->lock);
+	return FI_SUCCESS;
+}
 
-	progress_tx_queue(ep);
-	progress_rx_queue(ep);
+static void tcpx_auto_prog_tx_queues(struct tcpx_prog_mgr *mgr)
+{
+	struct dlist_entry *entry;
+	struct tcpx_ep *ep;
+	struct fid_list_entry *item;
+
+	fastlock_acquire(&mgr->lock);
+	dlist_foreach(&mgr->ep_list,entry) {
+		item = container_of(entry, struct fid_list_entry, entry);
+		ep = container_of(item->fid, struct tcpx_ep,
+				  util_ep.ep_fid.fid);
+		progress_tx_queue(ep);
+	}
+	fastlock_release(&mgr->lock);
+}
+
+void tcpx_manual_progress(struct util_ep *util_ep)
+{
+	struct tcpx_domain *domain;
+
+	domain = container_of(util_ep->domain,
+			      struct tcpx_domain,
+			      util_domain);
+
+	tcpx_auto_prog_tx_queues(&domain->progress.mgr);
+	tcpx_auto_prog_rx_queues(&domain->progress.mgr);
 	return;
 }
 
+static int tcpx_fid_match(struct dlist_entry *entry, const void *fid)
+{
+	struct fid_list_entry *item;
+
+	item = container_of(entry, struct fid_list_entry, entry);
+	return (item->fid == fid);
+}
+
+int tcpx_progress_ep_add(struct tcpx_progress *progress,
+			 struct tcpx_ep *ep)
+{
+	struct tcpx_prog_mgr *mgr;
+	struct dlist_entry *entry;
+	struct fid_list_entry *item;
+	void *new_ctxs;
+	uint64_t new_size;
+	int ret;
+
+	mgr = &progress->mgr;
+
+	fastlock_acquire(&mgr->lock);
+	entry = dlist_find_first_match(&mgr->ep_list, tcpx_fid_match,
+				       &ep->util_ep.ep_fid.fid);
+	if (entry) {
+		ret = FI_SUCCESS;
+		goto out;
+	}
+
+	item = calloc(1, sizeof(*item));
+	if (!item) {
+		ret = -FI_ENOMEM;
+		goto out;
+	}
+	item->fid = &ep->util_ep.ep_fid.fid;
+	dlist_insert_tail(&item->entry, &mgr->ep_list);
+
+	ret = fi_epoll_add(mgr->epoll_fd, ep->conn_fd, FI_EPOLL_IN,ep);
+	if (ret)
+		goto out1;
+
+	mgr->used++;
+	if (mgr->used > mgr->ctxs_sz) {
+		new_size = 2 * mgr->ctxs_sz;
+		new_ctxs = realloc(mgr->ctxs, new_size * sizeof(*mgr->ctxs));
+		if (!new_ctxs) {
+			mgr->ctxs = new_ctxs;
+			mgr->ctxs_sz = new_size;
+		}
+	}
+	fastlock_release(&mgr->lock);
+	return FI_SUCCESS;
+out1:
+	dlist_remove(&item->entry);
+	free(item);
+out:
+	fastlock_release(&mgr->lock);
+	return ret;
+}
+
+void tcpx_progress_ep_del(struct tcpx_progress *progress,
+			 struct tcpx_ep *ep)
+{
+	struct tcpx_prog_mgr *mgr;
+	struct dlist_entry *entry;
+	struct fid_list_entry *item;
+
+	mgr = &progress->mgr;
+
+	fastlock_acquire(&mgr->lock);
+	entry = dlist_remove_first_match(&mgr->ep_list, tcpx_fid_match,
+					 &ep->util_ep.ep_fid.fid);
+	if (!entry)
+		goto out;
+
+	item = container_of(entry, struct fid_list_entry, entry);
+	free(item);
+
+	fi_epoll_del(mgr->epoll_fd, ep->conn_fd);
+	mgr->used--;
+out:
+	fastlock_release(&mgr->lock);
+}
```

Comment:
ep_add allows inserting the same ep multiple times, but doesn't provide reference counting.  The result is that the first call to ep_del will remove the ep.

Reply:
added a reference count.